crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -254,46 +305,92 @@ mod tests {
         Ok(())
     }
 
+    fn generate_test_data(
+        db: Arc<AuthorityPerpetualTables>,
+        num_versions_per_object: u64,
+        num_object_versions_to_retain: u64,
+        total_unique_object_ids: u32,
+    ) -> Result<(Vec<ObjectKey>, Vec<ObjectKey>), anyhow::Error> {
+        let (mut to_keep, mut to_delete) = (vec![], vec![]);
+
+        let ids = ObjectID::in_range(ObjectID::ZERO, total_unique_object_ids.into())?;
+        for id in ids {
+            for (counter, i) in (0..num_versions_per_object).rev().enumerate() {
+                let object_key = ObjectKey(id, SequenceNumber::from_u64(i));
+                if counter < num_object_versions_to_retain.try_into().unwrap() {
+                    // latest `num_object_versions_to_retain` should not have been pruned
+                    to_keep.push(object_key);
+                } else {
+                    to_delete.push(object_key);
+                }
+                db.objects.insert(
+                    &ObjectKey(id, SequenceNumber::from(i)),
+                    &Object::immutable_with_id_for_testing(id),
+                )?;
+            }
+        }
+        assert_eq!(
+            to_keep.len() as u64,
+            std::cmp::min(num_object_versions_to_retain, num_versions_per_object)
+                * total_unique_object_ids as u64
+        );
+        Ok((to_keep, to_delete))
+    }
+
+    #[tokio::test]
+    async fn test_live_pruning() {
+        let path = tempfile::tempdir().unwrap().into_path();
+
+        let to_keep = {
+            let db = Arc::new(AuthorityPerpetualTables::open(&path, None));
+            let (to_keep, to_delete) = generate_test_data(db.clone(), 3, 2, 1000).unwrap();
+            let effects = TransactionEffects {
+                modified_at_versions: to_delete.into_iter().map(|o| (o.0, o.1)).collect(),
+                ..Default::default()
+            };
+            let checkpoint_state = CheckpointExecutionState {
+                effects: vec![effects],
+                checkpoint_sequence_number: 0,
+            };
+            let pruned =
+                AuthorityStorePruner::handle_checkpoint(checkpoint_state, &db.objects).unwrap();
+            assert_eq!(pruned, 1000);
```

Comment:
I think we could write a (sim)test which does owned and shared object transactions on a validator cluster and if everything is working as designed, we should have only one version of every object in the db eventually.  I can help with it.

Reply:
Great pointer, it'd make sense to add an object pruning test there. None of the workload triggers object deletion or unwrap like @mystenmark mentioned above, it'd be great to test that too(a different non sui-benchmarking test for this is also fine)

Reply:
live pruner is enabled by default, and load test modifies set of shared and owned objects. So live pruning is getting covered automatically (we'd get a failure if it deleted wrong version)